bindings/c/fdb_c.cpp

```diff
@@ -901,6 +911,41 @@ FDBFuture* fdb_transaction_get_range_impl(FDBTransaction* tr,
 	                    .extractPtr());
 }
 
+extern "C" DLLEXPORT FDBFuture* fdb_transaction_get_mapped_range_v2(FDBTransaction* tr,
```

Comment:
i spent ~ 2-3 weeks exploring this method before skipping it, i have almost zero confidence that it could work in all scenarios, especially when upgrades happens, reasons and questions below:
* reasons
1. the last time `FDB_API_CHANGED ` is used is for version `23`, and multi version transaction was not supported by that time.
2. the impl of `FDB_API_CHANGED` is so hacky, that I do not think we need to keep using it only because we had it before. I do not agree with your definition of `correct`.

* if you do want to have more discussion here, i have a few questions below
there are 2 cases we want to cover below, as you suggested
1. for the first case, it seems to be 7.1 Java -> 7.1 C -> 7.2 C talking to 7.2 cluster, how does it deal with the parameter in the step 7.1 C to 7.2 C?
2. for the second case, how does the 7.2 C binary(from primary client) forward the request to 7.1 C binary? so do we rely on the fact that shim will ignore the additional parameter from 7.2, and thus we can add new parameter to the end of the parameter list? 

In summary, it just sounds too fragile to me, it is super hard to write automated test against especially in the case of ongoing upgrades, and it could break any time due to compiler change, etc. I do not agree to try to have slightly cleaner code at the cost of robustness. please let me know, I can have an offline talk.
```
A 7.1 FDB Java with a 7.2 FDB C library talking to a 7.2 fdbserver.
A 7.2 FDB Java with a 7.2 FDB C library using the multi-version
client to talk to a 7.1 fdbserver via a second 7.1 FDB C library
```

Reply:
> the last time `FDB_API_CHANGED`  is used is for version 23, and multi version transaction was not supported by that time.

That's true, though it has been used to delete functions since then. I agree it's a shame that this sort of withered, though I think if I'm reading the code correctly, it should still work, even with the multi-version client. But it's true that there's not as much testing as we'd like here. I think I'd argue that that means we should have more tests with more versions of the client, but our testing of clients in general isn't that high

> the impl of `FDB_API_CHANGED` is so hacky, that I do not think we need to keep using it only because we had it before. I do not agree with your definition of correct.

I guess I'm not sure what definition of "correct" I'm using here, but my reasoning for wanting to use `FDB_API_CHANGED` is effectively two-fold. The first is that it is what we've done before, which is valuable because it is confusing if a codebase has multiple patterns for accomplishing the same thing. So, barring compelling reason otherwise, it would seem like a good idea to follow the pattern already established. (That being said, "it won't work with the multi-version client" could be compelling enough of a reason, though I'm not sure that it doesn't work.) The second reason is that the current approach has produced a lot of duplicated code and data structures, and using `FDB_API_CHANGED` should allow a lot of that to be collapsed into a single implementation plus some code bridging between the different versions. That being said, I think that that could also be done with two methods being exposed in the C API.

> for the first case, it seems to be 7.1 Java -> 7.1 C -> 7.2 C talking to 7.2 cluster, how does it deal with the parameter in the step 7.1 C to 7.2 C?

To be clear of what the proposal is, it is that the C API should have two methods, one of which is exactly like the current method (i.e., it doesn't have the extra parameters here and the data that get returned don't have the extra field), and the other of which is the new method. When the C API is loaded, it is given a header version, and if the header version >= 720, then the new method gets loaded; if the header version < 720, then the old function gets loaded.

When the 7.1 C client loads the 7.2 C client, it will call `select_api_version` with header version 710. (See: https://github.com/apple/foundationdb/blob/b227007ab0ce280606316d4ab7eb06457f9c3305/fdbclient/MultiVersionTransaction.actor.cpp#L1245 Note that the 7.1 C client will throw an error if a version > 710 is provided.) So the 7.2 C client will then ensure that the loaded value of `fdb_get_mapped_range` matches the signature expected by the 7.1 C client.

If the 7.1 Java bindings talk to a 7.2 C client directly, which they're allowed to do, then it will likewise pass 710 as the header version when selecting the API version, though it's maybe less obvious how. It calls the single-argument version of `select_api_version`: https://github.com/apple/foundationdb/blob/5bebb5b4aab810951d40a51cbabc014b73ba9bab/bindings/java/fdbJNI.cpp#L2259 Which then uses the `FDB_API_VERSION` constant as it's header version, which means it will be set to the latest version: https://github.com/apple/foundationdb/blob/5bebb5b4aab810951d40a51cbabc014b73ba9bab/bindings/c/foundationdb/fdb_c.h#L78 But in any case, this means that the `fdb_get_mapped_range` function loaded by the 7.1 FDB Java library will be the pre-720 version, will the parameters and return type that 7.1 FDB java expects.

> for the second case, how does the 7.2 C binary(from primary client) forward the request to 7.1 C binary? so do we rely on the fact that shim will ignore the additional parameter from 7.2, and thus we can add new parameter to the end of the parameter list?

In the case of a 7.2 client trying to use a 7.1 C binary, I believe the 7.1 C function should not be called with the extra parameter. In particular, when the `DLApi` struct is created, its `headerVersion` field is set during `selectApiVersion`. Then its `getMappedRange` method can use the value of that field to know whether it is safe to include the extra parameter and whether to expect the results of the function to include the new fields. There is a decision that needs to be made regarding whether it should reject requests to 7.1 that use the new parameters (with values that are incompatible with the 7.1 function), I suppose, but that's a bit of a different discussion.

> In summary, it just sounds too fragile to me, it is super hard to write automated test against especially in the case of ongoing upgrades, and it could break any time due to compiler change

It is a little fragile, perhaps. I don't think it's actually that hard to test, though. We'd need to have 7.2 and 7.1 fdbservers, maybe, and 7.2 and 7.1 fdb clients and validate that the code works at various versions. At one point, there might have been something in the bindingtester that was designed to try and test those scenarios, so the key point might be to make sure that that's done.

I don't think that compiler changes should affect this, though. The goal of the system is to not rely on things like the calling convention to ensure that extra parameters are safe (but ignored), but to instead have the right functions loaded at the right time based on the API version provided.

Reply:
I think this can be made to work with the MVC as well. We know what header version we are loading the external client at, and we can load the appropriate C function accordingly. There exists some similar precedent for executing different logic on the external clients at different API versions (see for example `DLApi::createDatabase`). I tend to agree that it would be better to follow the existing patterns for this rather than inventing new ones alongside the old ones, but I say that as an external observer without any particular vested interests.

Reply:
> for the second case, how does the 7.2 C binary(from primary client) forward the request to 7.1 C binary? so do we rely on the fact that shim will ignore the additional parameter from 7.2, and thus we can add new parameter to the end of the parameter list?

Probably the most principled way to handle this would be to not allow using a parameter that only exists at API version 720+ in earlier versions. If you want to use API version 720 and the new parameters, you would need to run with a stack that supported it all the way through (i.e. no-7.1 anywhere). I'm not sure this is really any different if you expose separate functions for the versioned APIs, since if you call the new one you still have to do some sort of translation or reject it when calling through to an older version.

Reply:
thanks for taking a look in detail. I totally agree that keep using FDB_API_CHANGED is a good idea, and I understand this current PR creates duplicate code, I also spent 2-3 weeks trying to get to the bottom of this, and I decided to have the PR as it is now for reasons:
* I do not see an easy way to write automated test, unlike a new API, I believe automated test is required if we are going toward this direction, I do not have confidence running this code in prod without automated test (normal scenarios plus the 2 special scenarios)
* one of the reason that automated test is required, is that there is no usage for this mechanism after multi-version client.
* There are too many subtleties, especially with multi-version client
That being said, if you believe we can have automated test to cover all scenarios and everything works, I am more than happy to have a meeting in the near future(allow me some time to re-visit the details before the meeting), and please advise how do we do that — especially the part that how does the shim work(primary C client talking to external C client).

Reply:
I plan to have a doc to enumerate all possible issues regarding different versions at different layers, (Java, primary C, external C, C++).  We can have a discussion to make sure there is no issue around them. It can also serve for future use. (I am still not sure how it works in the second scenario after reading you comments here, hopefully we can work that out)

Once that is done, i will see how to manually and automatically test them, and go on to the next step. @alecgrieser